uast/transformer/ast.go

```diff
@@ -0,0 +1,189 @@
+package transformer
+
+import (
+	"strings"
+
+	"gopkg.in/bblfsh/sdk.v1/uast"
+	"gopkg.in/bblfsh/sdk.v1/uast/role"
+)
+
+// SavePosOffset makes an operation that describes a uast.Position object with Offset field set to a named variable.
+func SavePosOffset(vr string) Op {
+	return TypedObj(uast.TypePosition, map[string]Op{
+		uast.KeyPosOff: Var(vr),
+	})
+}
+
+// SavePosLine makes an operation that describes a uast.Position object with Line field set to a named variable.
+func SavePosLine(vr string) Op {
+	return TypedObj(uast.TypePosition, map[string]Op{
+		uast.KeyPosLine: Var(vr),
+	})
+}
+
+// SavePosCol makes an operation that describes a uast.Position object with Col field set to a named variable.
+func SavePosCol(vr string) Op {
+	return TypedObj(uast.TypePosition, map[string]Op{
+		uast.KeyPosCol: Var(vr),
+	})
+}
+
+// Roles makes an operation that will check/construct a list of roles.
+func Roles(roles ...role.Role) ArrayOp {
+	arr := make([]Op, 0, len(roles))
+	for _, r := range roles {
+		arr = append(arr, Is(uast.String(r.String())))
+	}
+	return Arr(arr...)
+}
+
+// AppendRoles can be used to append more roles to an output of a specific operation.
+func AppendRoles(old ArrayOp, roles ...role.Role) ArrayOp {
+	if len(roles) == 0 {
+		return old
+	}
+	return AppendArr(old, Roles(roles...))
+}
+
+// ASTMap is a helper for creating a two-way mapping between AST and its normalized form.
+func ASTMap(name string, native, norm Op) Mapping {
+	return Mapping{
+		Name: name,
+		Steps: []Step{
+			{Name: "native", Op: native},
+			{Name: "norm", Op: norm},
+		},
+	}
+}
+
+// RolesField will create a roles field that appends provided roles to existing ones.
+// In case no roles are provided, it will save existing roles, if any.
+func RolesField(vr string, roles ...role.Role) Field {
+	return RolesFieldOp(vr, nil, roles...)
+}
+
+// RolesFieldOp is like RolesField but allows to specify custom roles op to use.
+func RolesFieldOp(vr string, op ArrayOp, roles ...role.Role) Field {
+	if len(roles) == 0 && op == nil {
+		return Field{
+			Name:     uast.KeyRoles,
+			Op:       Var(vr),
+			Optional: vr + "_exists",
+		}
+	}
+	var rop ArrayOp
+	if len(roles) != 0 && op != nil {
+		rop = AppendRoles(op, roles...)
+	} else if op != nil {
+		rop = op
+	} else {
+		rop = Roles(roles...)
+	}
+	return Field{
+		Name: uast.KeyRoles,
+		Op: If(vr+"_exists",
+			Append(NotEmpty(Var(vr)), rop),
+			rop,
+		),
+	}
+}
+
+// ASTObjectLeft construct a native AST shape for a given type name.
+func ASTObjectLeft(typ string, ast ObjectOp) ObjectOp {
```

Comment:
On both the go driver and the java one I've seen you add a `mapAST` and `mapASTCustom` wrapper that calls these. They should probably be in the SDK instead of each driver.

Also, `mapAST` almost always produce this annotation code:

```go
	mapAST("MethodInvocation", Obj{
		"expression":  OptObjectRoles("expr"),
		"name": OptObjectRoles("name"),
		"arguments": Each("args", ObjectRoles("arg")),
	}, Obj{
		"expression":  OptObjectRoles("expr", role.Call, role.Receiver),
		"name": OptObjectRoles("name", role.Call, role.Callee),
		"arguments": Each("args", ObjectRoles("arg", role.Call, role.Argument, role.Positional)),
	}, role.Expression, role.Call),
```

with the second object being exactly the same as the first one with the roles added, so a wrapper could just take the second Obj and infer the first part, which would help reduce the verbosity of the drivers.

Reply:
These functions exist in all drivers because I expected them to use the same approach as Go driver and mix additional roles automatically (like `Statement` or `Expression`) for every `mapAST` call depending on the type. But in practice, I inlined these roles for Java, and now doing the same for JS. So I agree, it makes sense to move it to SDK as well.

Regarding the second question, drivers also define `annotateType` that solves the problem that you described. However, it only covers simple cases where all field all pointing to a node or value. The case you describe here includes `Each` and thus it cannot be simplified using `annotateType` call.